### DISCUSSION TITLE
Is UpstreamConnectLayer deprecated ?!

### DISCUSSION THREAD

**Discussion by mrsiano** [mitmproxy/mitmproxy#4942] [General]
hi,
we've upgraded our mitm project and we can't import the following module:
from mitmproxy.proxy.layers.http import UpstreamConnectLayer

**Comment by mhils** [comment]
Yes, mitmproxy 7 has a completely new proxy core: https://mitmproxy.org/posts/releases/mitmproxy7/

**Reply by mrsiano** [reply]
another question,
previously we used the following code:
```
    assembled_request = http1.assemble_request(flow.request)
    flow.server_conn.write(assembled_request)
    flow.server_conn.wfile.flush()
    http1.read(
            rfile=flow.server_conn.rfile,
            request=flow.request,
            body_size_limit=body_size_limit
```

could you refer me to docs how to run it on mimproxy7 please ?

**Reply by mhils** [reply]
That's hard to answer without context on what you are actually trying to achieve here. The proxy core now operates in a sans-io mode (https://mitmproxy.org/posts/new-proxy-core/), so things are wildly different. Our upstream proxy logic is now handled in https://github.com/mitmproxy/mitmproxy/blob/main/mitmproxy/proxy/layers/http/_upstream_proxy.py.

**Reply by mrsiano** [reply]
first of all thanks,
the context is that we're trying to have ntlm authentication

**Reply by mrsiano** [reply]
using proxy on top of it

**Reply by mhils** [reply]
I don't know how NTLM authentication works, but if it does not require additional roundtrips modifying https://github.com/mitmproxy/mitmproxy/blob/main/mitmproxy/addons/upstream_auth.py should be good enough. Happy to accept a PR! Had you bothered to send one for the old proxy core, chances would be non-negligible that I would have ported it over. 😉

**Reply by mrsiano** [reply]
Thanks, so actually we need to keep using the same TCP connection for the nltm authentication session 
thats why we write the file previously 
```
    assembled_request = http1.assemble_request(flow.request)
    flow.server_conn.write(assembled_request)
    flow.server_conn.wfile.flush()
    http1.read(
            rfile=flow.server_conn.rfile,
            request=flow.request,
            body_size_limit=body_size_limit
```

can you think of any idea who we can use ntlm authenticaion with UpstreamProxy ?

**Reply by mhils** [reply]
I don't know how NTLM authentication works, so I can't really comment on this.

**Comment by mrsiano** [comment]
hi @mhils there is a way to evaluate another request on the same flow.server_conn of flow: http.HTTPFlow object ?

**Reply by mrsiano** [reply]
or another option is to get the http session() reference

**Comment by mrsiano** [comment]
@mhils ping
